openshift/default-catalog-consistency/pkg/extract/extract.go

```diff
@@ -0,0 +1,167 @@
+package extract
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	imgcopy "github.com/containers/image/v5/copy"
+	"github.com/containers/image/v5/docker"
+	"github.com/containers/image/v5/manifest"
+	"github.com/containers/image/v5/oci/layout"
+	"github.com/containers/image/v5/pkg/compression"
+	"github.com/containers/image/v5/signature"
+	"github.com/containers/image/v5/types"
+	"github.com/containers/storage/pkg/archive"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content/oci"
+)
+
+// Result represents the extracted OCI content and related context
+type Result struct {
+	Store   oras.ReadOnlyTarget
+	TmpDir  string
+	Cleanup func()
+}
+
+// UnpackImage pulls the image, extracts it to disk, and opens it as an OCI store.
+func UnpackImage(ctx context.Context, imageRef, name string) (res *Result, err error) {
+	tmpDir, err := os.MkdirTemp("", fmt.Sprintf("oci-%s-", name))
+	if err != nil {
+		return nil, fmt.Errorf("create temp dir: %w", err)
+	}
+
+	cleanup := func() { os.RemoveAll(tmpDir) }
+
+	if err := pullImage(ctx, imageRef, tmpDir); err != nil {
+		cleanup()
+		return nil, fmt.Errorf("pull image: %w", err)
+	}
+
+	if err := getLayers(ctx, tmpDir); err != nil {
+		cleanup()
+		return nil, fmt.Errorf("extract filesystem: %w", err)
+	}
+
+	store, err := oci.New(tmpDir)
+	if err != nil {
+		cleanup()
+		return nil, fmt.Errorf("open OCI store: %w", err)
+	}
+
+	return &Result{
+		Store:   store,
+		TmpDir:  tmpDir,
+		Cleanup: cleanup,
+	}, nil
+}
+
+// pullImage pulls an image and writes it to OCI layout at the specified directory
+func pullImage(ctx context.Context, imageRef, tmpDir string) error {
+	srcRef, err := docker.ParseReference("//" + imageRef)
+	if err != nil {
+		return fmt.Errorf("parse image ref: %w", err)
+	}
+
+	destRef, err := layout.ParseReference(fmt.Sprintf("%s:%s", tmpDir, "v1"))
+	if err != nil {
+		return fmt.Errorf("parse layout ref: %w", err)
+	}
+
+	// Force image resolution to Linux platform to avoid OS mismatch errors
+	// during testing on Mac. Example error:
+	// err: <*errors.errorString>{
+	//     s: "no image found in manifest list for architecture \"arm64\", variant \"v8\", OS \"darwin\"",
+	// }
+	sysCtx := &types.SystemContext{
+		OSChoice: "linux",
+	}
+
+	// On Linux systems with tools like podman, buildah, or skopeo, a default policy.json may
+	// be available at ~/.config/containers or /etc/containers.
+	// However, this is not the case on mac/docker envs.
+	// To avoid unnecessary setup complexity, we explicitly skip policy validation here only for local dev.
+	// Load default policy from standard locations (~/.config/containers/policy.json or /etc/containers/policy.json)
+	// If LOCAL_ENV=true, use a local insecure policy file
+	if os.Getenv("LOCAL_ENV") == "true" {
+		policyPath := filepath.Join("..", "..", "hack", "policy.json")
```

Comment:
is policy validation something that we'd also want to validate as part of these tests? e.g. to ensure they have been properly signed, etc....?

Reply:
Because I changed this one, and I think we need to discuss it, I will create a new comment with the code changed.